Add live vote register to motion/assignment-poll-detail

reiterl opened the pull request:
Resolve #3986 

First draft version.
Needs:
https://github.com/OpenSlides/openslides-autoupdate-service/pull/1118
https://github.com/OpenSlides/openslides-vote-service/pull/308

Reuse of the entitled-users-table.

MSoeb commented:
The first draft already looks good, but some changes are necessary:

1. layout - Live Voting Register
- Please use the existing column design with 'Single Votes' and 'Entitled users'. A separate view, as currently available, is not required. 
- Based on the previous point: The 'Single votes' column is NOT filled during an ongoing vote. The 'Entitled users' list only shows who has already voted.
- During active voting: The text 'No results yet.' should be replaced by the text 'Voting in progress.
- During active voting: A progress bar should be displayed below the text 'Voting in progress'. 
- During active voting: The 'Single votes' column should display a note that no data is available yet.  

2. data visualization:
- The 'Entitled users' list MUST include all persons called to vote. This includes all participants who are in the group(s) called to vote. This also implies participants who are not present.
- Vote delegations: Similar to the previous documentation and recording, transfers of voting rights should also be documented for nominal votes and be visible in the 'Entitled users' list already during the vote. The existing behavior can be reused here.

reiterl commented:
I chatted with @MSoeb and looked into the code.
I think the autoupdate service should provide a `entitled-users-live` field with the same data as entitled users at stop but live.

reiterl commented:
The current implementation uses the new `poll.has_voted_user_ids` field. This field provides a list of ids of user who already voted.
The live vote register should show all users, who are allowed to vote. (e.g. including users who haven't voted yet and delegates). The users who are allowed to vote depend on the `poll.entitled_groups` field and the delegates.
So there are some calculations needed and I am not sure how to handle the delegate to part.

reiterl commented:
I have worked on the different parts:

1.  all users who are allowed to vote are display.
2. delegation handling is also in place.
3. the live vote register into the is in the mat tab
4. in a running vote a new text is displayed
5. add a poll progress component

Please have a look on this.

luisa-beerboom commented on client/src/app/site/pages/meetings/modules/poll/base/base-poll-detail.component.ts:
```diff
@@ -276,6 +285,48 @@ export abstract class BasePollDetailComponent<V extends PollContentObject, S ext
+                        entries.push({
+                            id: user.id,
+                            user: user,
+                            voted_verbose: `voted:true`,
```
Is this supposed to always be `voted:true`?
If no, fix it. If yes, you may as well remove this property.

reiterl replied:
Fixed.

MSoeb reviewed (changes requested):
Some changes are required. 

1. Please use the existing column (tab) design with 'Single Votes' and 'Entitled users'. A separate view, as currently available, is not required. (relevant part in image: lower red square.)
2. Additionally more information is needed under the progress bar. Please add information about entitled users and present entitled users with numbers (see red marked area under the progress bar). Please take care, the numbers are not static and should change like the corresponding numbers does. The text should be right-alinged to the progress bar. The color should be gray, just like the words 'orange, schwarz' in the image. The information should be written in german in the following way:
- Stimmberechtige anwesend: [right number of progress bar]
- Stimmberechtigte gesamt: [number of entitled users]

![image](https://github.com/user-attachments/assets/2b880480-f3ff-4ca3-a11c-de98bfc358da)

MSoeb commented:
> To point 2. This is **not** part of this issue. I won't solve it. @MSoeb please open a new issue for drilling on the poll-progress.

Yes, this was not previously a requirement. However, this has changed. It is now a requirement. Please integrate it.

Explanation:
Without this new information, there is no explanation as to why the progress bar and the entitled voters list show a different number. In order to be able to understand what these different numbers mean without an explanation (e.g. in the FAQ), this information is absolutely essential. This was not previously foreseeable, but was added as a new, essential requirement for the feature during the last review.

Further information:
The information that is now to be added to the progress bar should always be displayed in the progress bar. For example, even when it is visible in the autopilot.

reiterl commented:
I think, the logic with the progress bar should be handled in another issue. I will exclude the poll-progress here, because it includes misleading information.

rrenkert commented:
@reiterl The additional text requested by @MSoeb is not about progress. It repeats the numbers of the progress bar and the list view to have an explanation what the numbers are about. These number should only be in sync with another.
Can you please integrate the additional text?

rrenkert reviewed (changes requested):
The progress was a required element in this view. please readd it.

reiterl commented:
I talked with @rrenkert about this missing two values. I will add them in the two views. @MSoeb no need for a new issue then.

MSoeb reviewed (changes requested):
Found different points, that needs to be fixed or improved:

1. By opening the entitled users tab (red square in image) during a live poll, an error appears in the console. -> Wanted behavior: Please fix error.

![image](https://github.com/user-attachments/assets/05d80804-1da5-4fd7-902d-c9091debccdf)

![image](https://github.com/user-attachments/assets/e889f466-416f-4af2-b82d-813dd76bfbed)

2. The numbers in the progress bar and in the information beneath it doesn't match.  (see examples in red squares). They don't match, because the lower number doesn't include vote delegations. -> Wanted behavior: Vote delegations should be included in the number 'present people allowed to vote. Both numbers should always match.

![image](https://github.com/user-attachments/assets/c725de25-7e2b-4d86-9c24-b38a9fb66aa9)

3. Re-Positioning of the new information (marked relevant content in image). Please place the information more next to the progress bar (marked with green) . It should have more space to the content below. 

![image](https://github.com/user-attachments/assets/18735558-4e4f-403c-9aec-bbdc55731221)

4. Re-Positioning of 'No data avilable'. The current position is not optimal. PLease please it, if possible, in the middle aligned to the word above. I've marked a possible area with green.

![image](https://github.com/user-attachments/assets/46965a1c-6a58-4fc0-bbb0-18f5ccce757f)

reiterl commented:
One other version of the the live vote register.
to point 1:  fixed the error
to point 2: Used the same way to select to get the present number as in the poll progress component.
to point 3: changed
to point 4: add a text align center and a color gray here ( as the global no-content class)

Please test.

MSoeb reviewed (changes requested):
Found two things, that needs to changed: 

1. Entitled users (during voting)
During a live voting the entitled users list (red square in image) isn't visible completely anymore. Only parts of it are visible. --> Whole list should be visible.
![image](https://github.com/user-attachments/assets/bee24073-367d-49c0-9d87-78822370985b)

2. Single votes list: position of 'no data available'
![image](https://github.com/user-attachments/assets/37dac611-2873-4154-93c8-cfb2fd210a9e)
Please add above and under 'No data available' space. Currently it doesn't look good. The element should be a bit more separated from the other elements.

reiterl commented:
To 2. added margin top/bottom to the 'no available data'.
To 1. In my tests the entitled user table component sometimes doesn't show all entries. It is not all the time and it seems to
 be a problem with this component. The observer had delivered the right data, as far as I can say. And it is not a problem with
an error in the entitled users html stuff. Most time I see the problem if I scroll. I will continue to debug here.

reiterl commented:
To point 1. In Scrolling table component there seems to be something wrong.
The itemSize and the item pixel buffer seems to be part of this problem. Changing the itemSize to 40 seems to fix the display problem. As adding more buffer, calculated from rowHeight.
Anyway in my test it works now. Please review and test.

luisa-beerboom commented on client/src/app/site/pages/meetings/pages/assignments/pages/assignment-polls/components/assignment-poll-detail/assignment-poll-detail.component.html:
```diff
@@ -63,6 +76,19 @@ <h1>{{ poll.title }}</h1>
+                                <os-entitled-users-table
+                                    [displayVoteWeight]="displayVoteWeight"
+                                    [entitledUsersObservable]="liveRegisterObservable"
+                                    [isViewingThis]="true"
```
I think `isViewingThis` is true by default and should never otherwise be set

MSoeb reviewed (approved):
looks good.
